src/stream_base.cc

```diff
@@ -100,94 +100,118 @@ int StreamBase::Writev(const FunctionCallbackInfo<Value>& args) {
+    // Determine storage size first
+    for (size_t i = 0; i < count; i++) {
+      storage_size = ROUND_UP(storage_size, WriteWrap::kAlignSize);
 
-  AsyncWrap* wrap = GetAsyncWrap();
-  // NOTE: All tests show that GetAsyncWrap() never returns nullptr here. If it
-  // can then replace the CHECK_NE() with if (wrap != nullptr).
-  CHECK_NE(wrap, nullptr);
-  env->set_init_trigger_id(wrap->get_id());
-  WriteWrap* req_wrap = WriteWrap::New(env,
-                                       req_wrap_obj,
-                                       this,
-                                       AfterWrite,
-                                       storage_size);
+      Local<Value> chunk = chunks->Get(i * 2);
 
-  uint32_t bytes = 0;
-  size_t offset = 0;
-  for (size_t i = 0; i < count; i++) {
-    Local<Value> chunk = chunks->Get(i * 2);
+      if (Buffer::HasInstance(chunk))
+        continue;
+        // Buffer chunk, no additional storage required
+
+      // String chunk
+      Local<String> string = chunk->ToString(env->isolate());
+      enum encoding encoding = ParseEncoding(env->isolate(),
+                                             chunks->Get(i * 2 + 1));
+      size_t chunk_size;
+      if (encoding == UTF8 && string->Length() > 65535)
+        chunk_size = StringBytes::Size(env->isolate(), string, encoding);
+      else
+        chunk_size = StringBytes::StorageSize(env->isolate(), string, encoding);
 
-    // Write buffer
-    if (Buffer::HasInstance(chunk)) {
+      storage_size += chunk_size;
+    }
+
+    if (storage_size > INT_MAX)
+      return UV_ENOBUFS;
+  } else {
+    for (size_t i = 0; i < count; i++) {
+      Local<Value> chunk = chunks->Get(i);
       bufs[i].base = Buffer::Data(chunk);
       bufs[i].len = Buffer::Length(chunk);
       bytes += bufs[i].len;
-      continue;
     }
 
-    // Write string
-    offset = ROUND_UP(offset, WriteWrap::kAlignSize);
-    CHECK_LE(offset, storage_size);
-    char* str_storage = req_wrap->Extra(offset);
-    size_t str_size = storage_size - offset;
-
-    Local<String> string = chunk->ToString(env->isolate());
-    enum encoding encoding = ParseEncoding(env->isolate(),
-                                           chunks->Get(i * 2 + 1));
-    str_size = StringBytes::Write(env->isolate(),
-                                  str_storage,
-                                  str_size,
-                                  string,
-                                  encoding);
-    bufs[i].base = str_storage;
-    bufs[i].len = str_size;
-    offset += str_size;
-    bytes += str_size;
+    // Try writing immediately without allocation
+    err = DoTryWrite(&buf_list, &count);
+    if (err != 0 || count == 0)
+      goto done;
   }
 
-  int err = DoWrite(req_wrap, *bufs, count, nullptr);
+  wrap = GetAsyncWrap();
+  // NOTE: All tests show that GetAsyncWrap() never returns nullptr here. If it
+  // can then replace the CHECK_NE() with if (wrap != nullptr).
+  CHECK_NE(wrap, nullptr);
```

Comment:
Heads up, this is going to conflict with https://github.com/nodejs/node/pull/13174 which removes the comment (because `GetAsyncWrap()` does actually never return a null pointer), so feel free to remove it here as well